docs/CONTRIBUTING.md

```diff
@@ -0,0 +1,108 @@
+
+## Formatting
+
+While we are still in the process of codifying code formats, whitespacing, etc, generally you can follow the precedents set around where you're making changes for information on how to indent lines, opening bracket positions, capitalization, variable naming, etc. Be specific, be as concisely elaborate/accurate as you can.  For example:
+
+```c++
+    if (CVarGetInteger("gCustomizeFrogsOcarinaGame", 0)) {
+        this->frogSongTimer = sTimerFrogSong[index] * CVarGetInteger("gFrogsModifyFailTime", 1);
+    } else {
+        this->frogSongTimer = sTimerFrogSong[index];
+    }
+```
+
+Here you see we open brackets on the same line, spacing outside the parentheses but not inside, same line else and opening bracket, etc. Also, all if...else blocks should be fully encapsulated with brackets, even if they're single-line blocks (which technically would be legal code).
+
+## Documentation
+
+We want to maintain the readability and integrity of vanilla codeflow as much as possible going forward. To do this, we have decided that we will want any modifications to vanilla code to be well notated and maintained separately from modified sections. Given this original code:
+
+```c++
+        this->frogSongTimer = 40;
+        this->ocarinaNoteIndex = 0;
+        func_8010BD58(play, OCARINA_ACTION_FROGS);
+        this->ocarinaNote = EnFr_GetNextNoteFrogSong(this->ocarinaNoteIndex);
+        EnFr_CheckOcarinaInputFrogSong(this->ocarinaNote);
+        this->actionFunc = EnFr_ContinueFrogSong;
+```
+
+If you wanted to modify this block with an enhancement to add difficulty options to the frog ocarina game, the first thing you would do is encapsulate it in a comment region as such:
+
+```c++
+        //Enhancement: Customize Frogs Ocarina Game
+        if (CVarGetInteger("gCustomizeFrogsOcarinaGame", 0) == 1) {
+            this->frogSongTimer = 40;
+            this->ocarinaNoteIndex = 0;
+            func_8010BD58(play, OCARINA_ACTION_FROGS);
+            this->ocarinaNote = EnFr_GetNextNoteFrogSong(this->ocarinaNoteIndex);
+            EnFr_CheckOcarinaInputFrogSong(this->ocarinaNote);
+            this->actionFunc = EnFr_ContinueFrogSong;
+        } else { // vanilla code
+            this->frogSongTimer = 40;
+            this->ocarinaNoteIndex = 0;
+            func_8010BD58(play, OCARINA_ACTION_FROGS);
+            this->ocarinaNote = EnFr_GetNextNoteFrogSong(this->ocarinaNoteIndex);
+            EnFr_CheckOcarinaInputFrogSong(this->ocarinaNote);
+            this->actionFunc = EnFr_ContinueFrogSong;
+        }
+        // Enhancement end
```

Comment:
this feels a little odd to me, in my mind the enhancement code ended when the `else` started

Reply:
The code did, but this is intended to denote the end of the section created by the separation of enhancement code and vanilla code, not just the enhancement-specific code.

Reply:
I realize now it might be a little redundant, but I wanted to err on the side of too much documentation rather than not enough.

Reply:
I've been thinking about what patterns would be best for this since I saw you make this PR. One that came to mind for small methods like `EnFr_SetupFrogSong` is

```c
void EnFr_SetupFrogSong_Original(EnFr* this, PlayState* play) {
    if (this->frogSongTimer != 0) {
        this->frogSongTimer--;
    } else {
        this->frogSongTimer = 40;
        this->ocarinaNoteIndex = 0;
        func_8010BD58(play, OCARINA_ACTION_FROGS);
        this->ocarinaNote = EnFr_GetNextNoteFrogSong(this->ocarinaNoteIndex);
        EnFr_CheckOcarinaInputFrogSong(this->ocarinaNote);
        this->actionFunc = EnFr_ContinueFrogSong;
    }
}

// Enhancement: Customize Frogs Ocarina Game (gCustomizeFrogsOcarinaGame)
void EnFr_SetupFrogSong_Enhanced(EnFr* this, PlayState* play) {
    if (this->frogSongTimer != 0) {
        this->frogSongTimer--;
    } else {
        this->frogSongTimer = 40 * CVarGetInteger("gFrogsModifyFailTime", 1);
        if (CVarGetInteger("gInstantFrogsGameWin", 0) == 1) {
            this->actor.textId = 0x40AC;
            EnFr_SetupReward(this, play, false);
        } else {
            this->ocarinaNoteIndex = 0;
            func_8010BD58(play, OCARINA_ACTION_FROGS);
            this->ocarinaNote = EnFr_GetNextNoteFrogSong(this->ocarinaNoteIndex);
            EnFr_CheckOcarinaInputFrogSong(this->ocarinaNote);
            this->actionFunc = EnFr_ContinueFrogSong;
        }
    }
}

void EnFr_SetupFrogSong(EnFr* this, PlayState* play) {
    if (CVarGetInteger("gCustomizeFrogsOcarinaGame", 0)) {
        EnFr_SetupFrogSong_Original(this, play);
    } else {
        EnFr_SetupFrogSong_Enhanced(this, play);
    }
}
```

Reply:
That is partially how Pepper0ni ended up doing his code for the fix for rando bean pricing and logic that he's working on, but he actually had a hybrid instead of completely duplicating outright, and I did like it. I just didn't know how much of the existing structure we wanted to maintain or not, and figured keeping things inline instead of pulling them out into their own function, would be preferable, at least for the vanilla blocks, and it didn't feel like the enhancement section was big enough to warrant pulling out into a separate function, though I know, eventually, Proxy wants to move in that direction (though it seemed to me like the idea was to put the custom functions in a separate file instead of side-by-side with the functions they're modifying or replacing, and I didn't think we were quite ready to follow that model).

Reply:
By the way, I'm only mentioning this to explain how I got where I did, not necessarily as an argument against your idea. I can definitely see the benefits to it, but it would be changing vanilla, technically, just by pulling vanilla code out into a new function.

Reply:
> it would be changing vanilla, technically, just by pulling vanilla code out into a new function

yeah, i can't think of a way to avoid that entirely

Reply:
> keeping things inline instead of pulling them out into their own function, would be preferable, at least for the vanilla blocks

i think i might feel the opposite. if the plan is to move all enhancement stuff out into it's own files, part of me thinks we should have
```
.
├── src
│   ├── boot
│   ├── buffers
│   ├── code
│   ├── dmadata
│   ├── elf_message
│   ├── libultra
│   └── overlays
├── src_authentic
│   ├── boot
│   ├── buffers
│   ├── code
│   ├── dmadata
│   ├── elf_message
│   ├── libultra
│   └── overlays
└── src_enhanced
    ├── boot
    ├── buffers
    ├── code
    ├── dmadata
    ├── elf_message
    ├── libultra
    └── overlays
```
where everything in `src` is just a wrapper method that starts out as
```
if (0) {
  // call enhanced version
} else {
  // call authentic version
}
```

but i know that gets messy when we start thinking about static vars etc.

overall i think it'd be good to have some conversations as a team about where we want to get to with this and possible paths towards it, then we can ensure whatever we recommend in this doc aligns with those goals/plans

Reply:
There’s a lot going on here, but I generally agree with the direction Bria is talking about.

A few points I want to add:
- This is the pattern I want to encourage for all code mods. Our “enhancements”, “randomizer”, “cheats”, and “dev tools” all being mods that are very easy to track down what all they change in the codebase.
- Function and variable names are preferable over comments explaining what’s ours and what’s not. Prime example of this is in the save struct, I have sectioned off the end of the struct for things we have added but that aren’t clearly named. Those fields should all either be prefixed with “soh” or “mod_name”, or be moved to their own struct that’s aptly named (a la sohStats).
- We don’t need to be DRY when it comes to the authentic code. This is code that by nature should never change anyway, so duplicating it into our own functions or files should be fine, the only thing we have to worry about is cross-mod compatibility, eg if we want to overwrite the same block of code/function for two different “mods” what’s the best way to handle that?

a lot of this will effect rando v3, as I’m planning on applying these types of changes to all of the actors we’ve touched for rando to clean them up and restore some original code paths.

Reply:
> if we want to overwrite the same block of code/function for two different “mods” what’s the best way to handle that?

in my "3 `src` directories" example i'd say
if the mods are mutually exclusive
```c
if (mod1cvar) {
  // call mod1 version
} else if (mod2cvar) {
  // call mod2 version
} else {
  // call authentic version
}
```

if the mods can coexist
```c
if (mod1cvar || mod2cvar) {
  // call modded version, handle multi mod complexity in there
} else {
  // call authentic version
}
```

Reply:
Man, there's a lot to unpack here. So, would that require converting all src files into wrappers, or would we only wrap things as they're edited or restored?

Reply:
I think it'd be better to have small changes be like the example but larger changes be put a c file somewhere else or separated in some way (as an example the more info on file select is big enough).
For stuff in the middle, adding functions to the original file but putting them at the top of the file that are clearly marked and separated from vanilla code would work.

Reply:
> would that require converting all src files into wrappers

that's what i was thinking. we could even have a 4th `src` dir and call it `src_soh_legacy` or `src_messy` and have the wrappers start by calling into that, then split up the work of cleaning up the mess (removing the file from `src_messy` and getting the changes working in the new pattern using only `src_authentic` and `src_enchanced`)